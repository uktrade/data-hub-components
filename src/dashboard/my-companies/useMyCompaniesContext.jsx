import { orderBy, pick } from 'lodash'
import { useReducer } from 'react'
import createUseContext from 'constate'
import { FILTER_CHANGE, LIST_CHANGE, ORDER_CHANGE } from './constants'

export const filterCompanyName = (companies, filterText) =>
  filterText.length
    ? companies.filter((c) =>
        c.company.name.toLowerCase().includes(filterText.toLowerCase())
      )
    : companies

const defaultState = {
  lists: [],
  selectedIdx: 0,
  sortBy: 'recent',
  filter: '',
}

export const reducer = (state, { type, ...action }) => {
  switch (type) {
    case LIST_CHANGE:
      return { ...state, selectedIdx: action.idx }
    case FILTER_CHANGE:
      return { ...state, filter: action.filter }
    case ORDER_CHANGE:
      return { ...state, sortBy: action.sortBy }
    default:
      return state
  }
}

// We are unpacking children here just to remove them from initialState.
const useMyCompaniesContext = createUseContext(
  ({ children, editListsLinkProps, ...rest }) => {
    const initialState = pick(rest, [
      'lists',
      'selectedIdx',
      'sortBy',
      'filter',
    ])
    const [state, dispatch] = useReducer(reducer, {
      ...defaultState,
      ...initialState,
      // Reducer assumes the list of company lists is sorted.
      lists: orderBy(initialState.lists, 'name'),
    })

    return {
      state,
      dispatch,
      editListsLinkProps,
    }
  }
)

export default useMyCompaniesContext
