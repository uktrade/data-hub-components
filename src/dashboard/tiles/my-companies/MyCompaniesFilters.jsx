import React, { useState } from 'react'

function MyCompaniesFilters({ initialCompanies, companies, onFilterUpdate }) {
  const [filterCompanyName, setFilterCompanyName] = useState('')

  function filterByName(e) {
    const companyName = e.target.value
    setFilterCompanyName(companyName)

    const filteredCompanies = initialCompanies.filter(
      c => c.companyName.toUpperCase().includes(companyName.toUpperCase()),
    )
    onFilterUpdate(filteredCompanies)
  }

  return (
    <div>
      <input type="text" onChange={filterByName} value={filterCompanyName} />
    </div>
  )
}

export default MyCompaniesFilters
