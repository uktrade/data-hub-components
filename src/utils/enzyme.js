export const withTargetValue = value => ({
  target: {
    value,
  },
})

/**
 * @function getOptionValueByText
 * @description Finds an `<option>` element with its inner text matching the
 * {text} parameter and returns the value of its `value` attribute.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {String} text - The option text
 */
export const getOptionValueByText = (wrapper, text) => (
  wrapper.find('option')
    .find({ children: text })
    .props()
    .value
)

/**
 * @function changeAndUpdate
 * @description Simulates the `change` event on the provided {wrapper} with
 * {vaule} as `target.value` and updates (re-renders) the wrapper.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {*} value - The value to be used as `event.target`
 */
export const changeAndUpdate = (wrapper, value) => {
  wrapper.simulate('change', withTargetValue(value))
  wrapper.update()
}

/**
 * @function changeSelectAndUpdate
 * @description Finds a `<select>` element in the wrapper, selects the option
 * matching the {text} and updates (re-renders) the wrapper.
 * @param {*} wrapper - Enzyme React wrapper
 * @param {String} text - The option text
 */
export const changeSelectAndUpdate = (wrapper, text) => {
  changeAndUpdate(wrapper.find('select'), getOptionValueByText(wrapper, text))
  wrapper.update()
}
