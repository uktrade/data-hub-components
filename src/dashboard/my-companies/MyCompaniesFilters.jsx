import React from 'react'
import styled from 'styled-components'
import { SelectInput } from '@govuk-react/select'
import LabelText from '@govuk-react/label-text'
import { MEDIA_QUERIES, SPACING } from '@govuk-react/constants'
import { typography } from '@govuk-react/lib'
import useMyCompaniesContext from './useMyCompaniesContext'
import { FILTER_CHANGE, ORDER_CHANGE } from './constants'

const StyledSelectInput = styled(SelectInput)(typography.font({ size: 14 }), {
  [MEDIA_QUERIES.LARGESCREEN]: {
    width: 'auto',
  },
})

const StyledLabelText = styled(LabelText)(typography.font({ size: 14 }))

const StyledContainer = styled.div`
  display: flex;
  justify-content: space-between;
`

const StyledSearch = styled.input`
  margin-right: ${SPACING.SCALE_2};
  padding-right: ${SPACING.SCALE_1};
  padding-left: ${SPACING.SCALE_1};
`

function MyCompaniesFilters() {
  const { dispatch } = useMyCompaniesContext()

  return (
    <StyledContainer>
      <StyledSearch
        type="text"
        placeholder="Search list"
        onChange={e => dispatch({
          type: FILTER_CHANGE,
          filter: e.target.value,
        })}
      />
      <StyledLabelText>
        Sort By:{' '}
        <StyledSelectInput
          onChange={e => dispatch({
            type: ORDER_CHANGE,
            sortBy: e.target.value,
          })}
        >
          <option value="recent">Recent interaction</option>
          <option value="least-recent">Least recent interaction</option>
          <option value="alphabetical">Company name A-Z</option>
        </StyledSelectInput>
      </StyledLabelText>
    </StyledContainer>
  )
}

export default MyCompaniesFilters
