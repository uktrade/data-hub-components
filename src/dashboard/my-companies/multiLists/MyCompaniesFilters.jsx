import React, { useRef } from 'react'
import styled from 'styled-components'
import { SelectInput } from '@govuk-react/select'
import LabelText from '@govuk-react/label-text'
import { MEDIA_QUERIES, SPACING } from '@govuk-react/constants'
import { typography } from '@govuk-react/lib'
import ACTIONS from '../constants'
import useMyCompaniesContext from '../useMycompaniesListsContext'

const StyledSelectInput = styled(SelectInput)(typography.font({ size: 14 }), {
  [MEDIA_QUERIES.LARGESCREEN]: {
    width: 'auto',
  },
})

const StyledLabelText = styled(LabelText)(typography.font({ size: 14 }))

const StyledContainer = styled.div`
  display: flex;
  justify-content: space-between;
`

const StyledSearch = styled.input`
  margin-right: ${SPACING.SCALE_2};
  padding-right: ${SPACING.SCALE_1};
  padding-left: ${SPACING.SCALE_1};
`

function MyCompaniesFilters() {
  const { state, dispatch } = useMyCompaniesContext()
  const inputEl = useRef()

  const handleSelectChange = (e) => {
    dispatch({ type: ACTIONS.SORT_BY, sortType: e.target.value })
  }

  const handleInputChange = (e) => {
    dispatch({ type: ACTIONS.FILTER_BY, filterText: e.target.value })
  }

  return (
    <StyledContainer>
      <StyledSearch
        type="text"
        placeholder="Search list"
        onChange={handleInputChange}
        value={state.filterText}
        ref={inputEl}
      />
      <StyledLabelText>
        Sort By:{' '}
        <StyledSelectInput onChange={handleSelectChange} value={state.sortType}>
          <option value="recent">Recent interaction</option>
          <option value="least-recent">Least recent interaction</option>
          <option value="alphabetical">Company name A-Z</option>
        </StyledSelectInput>
      </StyledLabelText>
    </StyledContainer>
  )
}

export default MyCompaniesFilters
