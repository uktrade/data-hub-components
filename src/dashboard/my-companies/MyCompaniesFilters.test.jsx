import React from 'react'
import { mount } from 'enzyme'
import MyCompaniesFilters from './MyCompaniesFilters'
import useMyCompaniesContext from './useMyCompaniesContext'

describe('MyCompaniesFilters', () => {
  const mockDispatch = jest.fn()
  test('renders company filters', () => {
    const wrapper = mount(
      <useMyCompaniesContext.Provider>
        <MyCompaniesFilters />
      </useMyCompaniesContext.Provider>,
    )
    expect(wrapper.debug()).toMatchSnapshot()
  })

  describe('Select', () => {
    test('should call store actions when event handler fires', () => {
      const wrapper = mount(
        <useMyCompaniesContext.Provider mockProps={{ dispatch: mockDispatch }}>
          <MyCompaniesFilters />
        </useMyCompaniesContext.Provider>,
      )
      wrapper.find('select').simulate('change', { target: { value: 'recent' } })

      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'sortBy',
        sortType: 'recent',
      })
    })
  })

  describe('Search', () => {
    test('should call store actions when event handler fires', () => {
      const wrapper = mount(
        <useMyCompaniesContext.Provider mockProps={{ dispatch: mockDispatch }}>
          <MyCompaniesFilters />
        </useMyCompaniesContext.Provider>,
      )
      wrapper.find('input').simulate('change', { target: { value: 'hello' } })

      expect(mockDispatch).toHaveBeenCalledWith({
        type: 'filterBy',
        filterText: 'hello',
      })
    })
  })
})
