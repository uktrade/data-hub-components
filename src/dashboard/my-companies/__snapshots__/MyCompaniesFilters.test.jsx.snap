// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyCompaniesFilters renders company filters 1`] = `
"<Provider>
  <MyCompaniesFilters>
    <styled.div>
      <StyledComponent forwardedComponent={{...}} forwardedRef={{...}}>
        <div className=\\"sc-gzVnrw ksvrPZ\\">
          <styled.input type=\\"text\\" placeholder=\\"Search list\\" onChange={[Function: handleInputChange]} value=\\"\\">
            <StyledComponent type=\\"text\\" placeholder=\\"Search list\\" onChange={[Function: handleInputChange]} value=\\"\\" forwardedComponent={{...}} forwardedRef={{...}}>
              <input type=\\"text\\" placeholder=\\"Search list\\" onChange={[Function: handleInputChange]} value=\\"\\" className=\\"sc-htoDjs kyDyMB\\" />
            </StyledComponent>
          </styled.input>
          <Styled(LabelText)>
            <StyledComponent forwardedComponent={{...}} forwardedRef={{...}}>
              <LabelText className=\\"sc-bZQynM kxBMOH\\">
                <styled.span className=\\"sc-bZQynM kxBMOH\\">
                  <StyledComponent className=\\"sc-bZQynM kxBMOH\\" forwardedComponent={{...}} forwardedRef={{...}}>
                    <span className=\\"sc-bZQynM kxBMOH sc-bwzfXH lbxcXc\\">
                      Sort By:
                       
                      <Styled(styled.select) onChange={[Function: handleSelectChange]} value=\\"recent\\">
                        <StyledComponent onChange={[Function: handleSelectChange]} value=\\"recent\\" forwardedComponent={{...}} forwardedRef={{...}}>
                          <select onChange={[Function: handleSelectChange]} value=\\"recent\\" className=\\"sc-ifAKCX sc-EHOje aZems\\">
                            <option value=\\"recent\\">
                              Recent interaction
                            </option>
                            <option value=\\"least-recent\\">
                              Least recent interaction
                            </option>
                            <option value=\\"alphabetical\\">
                              Company name A-Z
                            </option>
                          </select>
                        </StyledComponent>
                      </Styled(styled.select)>
                    </span>
                  </StyledComponent>
                </styled.span>
              </LabelText>
            </StyledComponent>
          </Styled(LabelText)>
        </div>
      </StyledComponent>
    </styled.div>
  </MyCompaniesFilters>
</Provider>"
`;
